python/unthrow/unthrow/unthrow.pyx

```diff
@@ -0,0 +1,648 @@
+                                   # doesn't redeclare it so we can just put 1 in
+        int f_iblock
+        int f_trace_opcodes;
+
+
+    ctypedef _frame PyFrameObject
+
+    cdef PyFrameObject* PyEval_GetFrame()
+    cdef void PyFrame_FastToLocals(PyFrameObject* frame)
+    cdef void PyFrame_LocalsToFast(PyFrameObject* frame,int clear)
+    cdef PyObject* PyObject_Call(PyObject*obj,PyObject*args,PyObject*kwArgs)
+
+
+cdef extern from "opcode.h":
+    IF PythonVersion==3.8:
+        cdef enum _opcodes:
+            SETUP_FINALLY
+            BEGIN_FINALLY
+    ELSE:
+        # 3.9 or newer
+        cdef enum _opcodes:
+            SETUP_FINALLY
+            DUP_TOP
+            POP_TOP
+            SETUP_WITH
+            SETUP_ASYNC_WITH
+
+
+cdef extern from "Python.h":
+    ctypedef struct PyCodeObject:
+        PyObject* co_code
+        PyObject* co_name
+        PyObject* co_filename
+        int co_flags
+
+    cdef PyObject* Py_True
+    ctypedef int (*Py_tracefunc)(PyObject *obj, PyFrameObject *frame, int what, PyObject *arg)
+    cdef enum _trace_types:
+        PyTrace_LINE
+        PyTrace_CALL
+        PyTrace_C_CALL
+        PyTrace_OPCODE
+        PyTrace_RETURN
+
+    cdef int PyCompile_OpcodeStackEffectWithJump(int,int,int)
+    cdef char* PyBytes_AsString(PyObject *o)
+    cdef PyObject* PyBytes_FromStringAndSize(const char *v, Py_ssize_t len)
+    cdef Py_ssize_t PyObject_Length(PyObject *o)
+    cdef PyObject* PyDict_GetItem(PyObject*,PyObject*)
+    cdef int PyDict_SetItemString(PyObject*dict,const char*key,PyObject* val)
+    cdef int PyDict_SetItem(PyObject*dict,PyObject*key,PyObject* val)
+    cdef int PyDict_Next(PyObject *p, Py_ssize_t *ppos, PyObject **pkey, PyObject **pvalue)
+
+    cdef int PyCode_Check(PyObject*)
+
+    cdef int PyEval_SetTrace(Py_tracefunc fn,PyObject*self)
+    cdef int Py_AddPendingCall(int (*fn)(void*),void*arg)
+    cdef void PyErr_SetObject(PyObject*type,PyObject*val)
+    cdef PyObject* PyUnicode_FromString(char*str);
+    cdef PyObject* PyExc_ValueError
+
+    cdef PyObject* PyDict_New()
+
+    cdef void PyErr_SetString(PyObject*,char *)
+    cdef PyObject* PyErr_NewException(const char *name, PyObject *base, PyObject *dict)
+    cdef PyObject* PyErr_Occurred()
+    cdef PyObject* PyTuple_New(Py_ssize_t size)
+
+    cdef int Py_REFCNT(PyObject* ob)
+
+    cdef PyObject* PyList_GetItem(PyObject*,Py_ssize_t pos)
+
+    cdef PyObject* PyObject_Type(PyObject*ob)
+    cdef int PyGen_Check(PyObject* ob)
+    ctypedef struct PyGenObject:
+        PyFrameObject *gi_frame
+
+cdef PyObject* ResumableExceptionClass
+
+# ResumableExceptionClass=<PyObject*>PyErr_NewException("unthrow.ResumableException",NULL,NULL)
+# ResumableException=<object>ResumableExceptionClass
+
+class ResumableException(BaseException):
+    pass
+
+ResumableExceptionClass = <PyObject*>ResumableException
+
+cdef _get_stack_pos(object code,int target,int before):
```

Comment:
This is the tricky part:
https://github.com/a-rahimi/python-checkpointing#failed-attempts

https://stackoverflow.com/questions/44346433/in-c-python-accessing-the-bytecode-evaluation-stack/44443331#44443331

Basically, f_stacktop won't be set unless it's in trace or generator/coroutine and thus, it's hard to get the stack size.

The good thing is that now `dis` can tell you how the stack is going to be affected for each opcode. (https://github.com/a-rahimi/python-checkpointing2/issues/2#issue-821890262)